store/json.go

```diff
@@ -55,9 +57,115 @@ func NewJsonFileStore(fileName string) (KeyValueStore, error) {
 	}
 
 	kvs := &jsonFileStore{
-		fileName: fileName,
+		fileName:     fileName,
 		lockFileName: platform.CNILockPath + filepath.Base(fileName) + lockExtension,
-		data:     make(map[string]*json.RawMessage),
+		data:         make(map[string]*json.RawMessage),
+	}
+
+	return kvs, nil
+}
+
+// fileExists states the passed filename and returns if it exists.
+// if any error except for os.ErrNotExists is encountered, it is returned.
+func fileExists(filename string) (bool, error) {
+	if _, err := os.Stat(filename); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// NewJsonFileStore creates a new jsonFileStore object, accessed as a KeyValueStore.
+func NewJsonFileStoreOrMigrate(legacyFileName, fileName string) (KeyValueStore, error) {
+	if fileName == "" {
+		fileName = defaultFileName
+	}
+
+	if platform.CNILockPath != "" {
+		err := os.MkdirAll(platform.CNILockPath, os.FileMode(0664))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create the state subdirectory
+	if platform.CNIRuntimePath != "" {
+		err := os.MkdirAll(platform.CNIRuntimePath, os.FileMode(0644))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create the state store
+	kvs := &jsonFileStore{
+		fileName:     fileName,
+		lockFileName: platform.CNILockPath + filepath.Base(fileName) + lockExtension,
+		data:         make(map[string]*json.RawMessage),
+	}
+
+	// test if the backing state file exists - if so, return
+	ok, err := fileExists(fileName)
+	if err != nil {
+		return nil, err
+	}
+	if ok {
+		return kvs, nil
+	}
+
+	// if we don't have a legacy path, we won't attempt to do any of the
+	// below path migrations.
+	if legacyFileName == "" {
+		return kvs, nil
+	}
+
+	// the statefile does not exist, so we need to lock and check if there is
+	// a legacy statefile that we need to migrate
+	if err := kvs.Lock(true); err != nil {
+		return nil, err
+	}
+	defer kvs.Unlock(false)
+
+	// check if the statefile exists again, now that we have acquired the lock,
+	// just in case someone else wrote it since we checked but before we locked
+	ok, err = fileExists(fileName)
```

Comment:
potentially wrap the error with context, calling `fileExists(fileName)` multiple times in this func on the same filepath and returning the error could be difficult to trace